beacon-chain/rpc/prysm/beacon/ssz_query.go

```diff
@@ -0,0 +1,228 @@
+	"github.com/OffchainLabs/prysm/v6/beacon-chain/rpc/lookup"
+	"github.com/OffchainLabs/prysm/v6/encoding/ssz/query"
+	"github.com/OffchainLabs/prysm/v6/monitoring/tracing/trace"
+	"github.com/OffchainLabs/prysm/v6/network/httputil"
+	ethpb "github.com/OffchainLabs/prysm/v6/proto/prysm/v1alpha1"
+	sszquerypb "github.com/OffchainLabs/prysm/v6/proto/ssz_query"
+	"github.com/OffchainLabs/prysm/v6/runtime/version"
+)
+
+// QueryBeaconState handles SSZ Query request for BeaconState.
+// Returns as bytes serialized SSZQueryResponse.
+func (s *Server) QueryBeaconState(w http.ResponseWriter, r *http.Request) {
+	ctx, span := trace.StartSpan(r.Context(), "beacon.QueryBeaconState")
+	defer span.End()
+
+	stateID := r.PathValue("state_id")
+	if stateID == "" {
+		httputil.HandleError(w, "state_id is required in URL params", http.StatusBadRequest)
+		return
+	}
+
+	// Validate path before lookup: it might be expensive.
+	var req structs.QuerySSZRequest
+	err := json.NewDecoder(r.Body).Decode(&req)
+	switch {
+	case errors.Is(err, io.EOF):
+		httputil.HandleError(w, "No data submitted", http.StatusBadRequest)
+		return
+	case err != nil:
+		httputil.HandleError(w, "Could not decode request body: "+err.Error(), http.StatusBadRequest)
+		return
+	}
+
+	if len(req.Query) == 0 {
+		httputil.HandleError(w, "Empty query submitted", http.StatusBadRequest)
+		return
+	}
+
+	path, err := query.ParsePath(req.Query)
+	if err != nil {
+		httputil.HandleError(w, "Could not parse path '"+req.Query+"': "+err.Error(), http.StatusBadRequest)
+		return
+	}
+
+	stateRoot, err := s.Stater.StateRoot(ctx, []byte(stateID))
+	if err != nil {
+		var rootNotFoundErr *lookup.StateRootNotFoundError
+		if errors.As(err, &rootNotFoundErr) {
+			httputil.HandleError(w, "State root not found: "+rootNotFoundErr.Error(), http.StatusNotFound)
+			return
+		}
+		httputil.HandleError(w, "Could not get state root: "+err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	st, err := s.Stater.State(ctx, []byte(stateID))
+	if err != nil {
+		shared.WriteStateFetchError(w, err)
+		return
+	}
+
+	var sszObject query.SSZObject
+	switch st.Version() {
+	case version.Phase0:
+		sszObject = st.ToProto().(*ethpb.BeaconState)
+	case version.Altair:
+		sszObject = st.ToProto().(*ethpb.BeaconStateAltair)
+	case version.Bellatrix:
+		sszObject = st.ToProto().(*ethpb.BeaconStateBellatrix)
+	case version.Capella:
+		sszObject = st.ToProto().(*ethpb.BeaconStateCapella)
+	case version.Deneb:
+		sszObject = st.ToProto().(*ethpb.BeaconStateDeneb)
+	case version.Fulu:
+		sszObject = st.ToProto().(*ethpb.BeaconStateFulu)
+	default:
+		httputil.HandleError(w, "Unsupported state version for querying: "+version.String(st.Version()), http.StatusBadRequest)
+		return
+	}
+
+	info, err := query.AnalyzeObject(sszObject)
+	if err != nil {
+		httputil.HandleError(w, "Could not analyze state object: "+err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	encodedState, err := st.MarshalSSZ()
```

Comment:
This can be moved to after `CalculateOffsetAndLength` because there's no point in marshaling if that function fails.